transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy

```diff
@@ -237,101 +240,149 @@ class HibernateCriteriaQueryBuilder implements QueryBuilder<Criterion, DetachedC
+                    criterion = Restrictions.in(propertyName, value)
                 }
+                break
             case Operator.LIKE:
-                return StringUtils.like(propertyName, constraint.value.toString(), MatchMode.EXACT)
+                criterion = StringUtils.like(propertyName, value.toString(), MatchMode.EXACT)
+                break
             case Operator.IN:
-                return Restrictions.in(propertyName, constraint.value)
+                criterion = Restrictions.in(propertyName, value)
+                break
             default:
-                throw new QueryBuilderException("Operator '${constraint.operator.symbol}' not supported.")
+                throw new QueryBuilderException("Operator '${operator.symbol}' not supported.")
+        }
+        if (type == Type.DATE) {
+            Restrictions.and(
+                Restrictions.isNotNull(propertyName),
+                Restrictions.ne(propertyName, EMPTY_DATE),
+                criterion
+            )
+        } else {
+            criterion
+        }
+    }
+
+    /**
+     * Creates a criteria object for a field constraint. Applies {@link #convertValue(Field, Object)} on the value
+     *
+     * @throws QueryBuilderException if the field type does not support the operator or the value is not supported
+     * for the field type.
+     * @see {@link Operator} and {@link Type} for supported operators and types.
+     */
+    Criterion build(FieldConstraint constraint) {
+        assert constraint.field != null
+        if (!constraint.operator.supportsType(constraint.field.type)) {
+            throw new QueryBuilderException("Field type ${constraint.field.type} not supported for operator '${constraint.operator.symbol}'.")
+        }
+        if ((!(constraint.operator in [Operator.BETWEEN, Operator.IN]) && !constraint.field.type.supportsValue(constraint.value))
+                || (constraint.operator in [Operator.BETWEEN, Operator.IN] && !constraint.field.type.supportsValue(constraint.value.first()))) {
+            throw new QueryBuilderException("Value of class ${constraint.value?.class?.simpleName} not supported for field type '${constraint.field.type}'.")
+        }
+        constraint.value = convertValue(constraint.field, constraint.value)
+        if (constraint.field.dimension == VisitDimension) {
+            /**
+             * special case that requires a subquery, because there is no proper
+             * reference to the visit dimension in {@link ObservationFact}.
+             */
+            DetachedCriteria subCriteria = DetachedCriteria.forClass(org.transmartproject.db.i2b2data.VisitDimension, 'visit')
+            String propertyName = constraint.field.fieldName
+            subCriteria.add(applyOperator(constraint.operator, propertyName, constraint.field.type, constraint.value))
+            return Subqueries.propertiesIn(['encounterNum', 'patient'] as String[],
```

Comment:
Doesn't groovy do this `String[]` conversion automatically to match the value to the parameter type?